vllm/v1/worker/gpu_model_runner.py

```diff
@@ -995,6 +1002,150 @@ def _update_states_after_model_execute(
         for i, num_tokens in enumerate(num_accepted_tokens):
             self.input_batch.num_accepted_tokens_cpu[i] = num_tokens
 
+    def _update_ngram_gpu_tensors(self, scheduler_output: "SchedulerOutput") -> None:
+        """Incrementally update token_ids_gpu_tensor and num_tokens_no_spec_gpu
+        for ngram GPU proposer to avoid redundant CPU-GPU transfers.
+
+        This follows a similar pattern to _prepare_input_ids for efficient
+        batch updates when requests change between iterations.
+        """
+        prev_req_id_to_index = self.input_batch.prev_req_id_to_index
+        curr_req_id_to_index = self.input_batch.req_id_to_index
+
+        # If no previous batch or batch is empty, initialize all from scratch
+        if prev_req_id_to_index is None or not curr_req_id_to_index:
+            if curr_req_id_to_index:
+                # Initialize all token_ids from requests
+                for req_id, idx in curr_req_id_to_index.items():
+                    req_state = self.requests[req_id]
+                    # Get prompt_token_ids + output_token_ids
+                    prompt_token_ids = (
+                        req_state.prompt_token_ids
+                        if req_state.prompt_token_ids is not None
+                        else []
+                    )
+                    all_token_ids = prompt_token_ids + req_state.output_token_ids
+                    num_tokens = len(all_token_ids)
+                    # Copy to GPU tensor
+                    self.input_batch.token_ids_gpu_tensor[idx, :num_tokens].copy_(
+                        torch.tensor(
+                            all_token_ids, dtype=torch.int32, device=self.device
+                        ),
+                        non_blocking=True,
+                    )
+                    self.input_batch.num_tokens_no_spec_gpu[idx] = num_tokens
```

Comment:
Can this logic be integrated into `_update_states`, where `num_tokens_no_spec` (cpu) is maintained? That seems like it would be cleaner than recomputing it twice and copying over here. Also, we would not want to maintain two pieces of the same logic.